File: 백준/1717/solution2.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

# 특정 원소가 속한 집합을 찾는 함수
def find(parent, x):
    if parent[x] == x:
        return x
    return find(parent, parent[x])

# 두 원소가 속한 집합을 합치기
def union(parent, a, b):
    aParent = find(parent, a)
    bParent = find(parent, b)
    if aParent < bParent:
        parent[bParent] = aParent
    else:
        parent[aParent] = bParent

N, M = map(int, input().split())
# 초기화
parent = [0] * (N+1)

for i in range(N+1):
    parent[i] = i

# inputs
for _ in range(M):
    cmd, a, b = map(int, input().split())
    if cmd == 0:
        #union
        if a != b:
            union(parent, a, b)
    else:
        # find
        if a == b or find(parent, a) == find(parent, b):
            print('YES')
        else:
            print('NO')

File: 백준/1976/solution2.swift
import Foundation

func find(_ parent: inout [Int], _ node: Int) -> Int {
    if parent[node] != node {
        parent[node] = find(&parent, parent[node])
    }
    return parent[node]
}

func union(_ parent: inout [Int], _ a: Int, _ b: Int) {
    let aParent = find(&parent, a)
    let bParent = find(&parent, b)

    if aParent < bParent {
        parent[bParent] = aParent
    } else {
        parent[aParent] = bParent
    }
}

let N = Int(readLine()!)!
let M = Int(readLine()!)!

var parent: [Int] = []

for i in 0..<N {
    parent.append(i)
}

for i in 0..<N {
    let data = readLine()!.split(separator: " ").compactMap { Int($0) }

    for (j, isPath) in data.enumerated() {
        if isPath == 1 {
            union(&parent, i, j)
        }
    }
}
let travelPath = readLine()!.split(separator: " ").compactMap { Int($0)! - 1 }

for i in 1..<travelPath.count {
    let start = travelPath[i-1]
    let destination = travelPath[i]
    
    if find(&parent, start) != find(&parent, destination) {
        print("NO")
        exit(0)
    }
}
print("YES")